assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx

```diff
@@ -0,0 +1,257 @@
+ *    "21": "Decor"
+ * }
+ * In the example above, "category1" is a duplicated term name, so the term's slug is appended for distinction.
+ *
+ * termNameToIdMap is the inverse of termIdToNameMap, mapping term names back to their respective IDs.
+ */
+const useTermMaps = (
+	taxonomy: Taxonomy
+): {
+	termIdToNameMap: Map< number, string >;
+	termNameToIdMap: Map< string, number >;
+	isResolving: boolean;
+} => {
+	// Fetch all terms for the given taxonomy.
+	const { records: allTerms, isResolving: isResolvingAllTerms } =
+		useEntityRecords< Term[] >( 'taxonomy', taxonomy.slug, {
+			...BASE_QUERY_ARGS,
+		} );
+
+	// Memoize the result to avoid re-renders.
+	return useMemo( () => {
+		const termIdToNameMap = new Map< number, string >();
+		const termNameToIdMap = new Map< string, number >();
+
+		if ( ! allTerms )
+			return {
+				termIdToNameMap,
+				termNameToIdMap,
+				isResolving: isResolvingAllTerms,
+			};
+
+		// Count the number of times a term name appears.
+		const nameCountMap = createNameCountMap( allTerms );
+
+		// Create the map with term ids as keys and term names as values.
+		for ( const term of allTerms ) {
+			const termId = term.id;
+			const termName = term.name;
+			const name = generateUniqueName(
+				termName,
+				term.slug,
+				nameCountMap
+			);
+			termIdToNameMap.set( termId, name );
+			termNameToIdMap.set( name, termId );
+		}
+		return {
+			termIdToNameMap,
+			termNameToIdMap,
+			isResolving: isResolvingAllTerms,
+		};
+	}, [ allTerms, isResolvingAllTerms ] );
+};
+
+const TaxonomyItem = ( { taxonomy, termIds, onChange }: TaxonomyItemProps ) => {
+	const [ search, setSearch ] = useState( '' );
+	const suggestionsRef = useRef< string[] >( EMPTY_ARRAY );
+	const currentValueRef = useRef<
+		{
+			id: number;
+			value: string;
+		}[]
+	>( EMPTY_ARRAY );
+
+	// Search is debounced to limit the number of API calls as the user types
+	const debouncedSearch = useDebounce( setSearch, 250 );
+
+	const {
+		termIdToNameMap,
+		termNameToIdMap,
+		isResolving: isResolvingTermMaps,
+	} = useTermMaps( taxonomy );
+
+	// Fetch the terms based on the search query.
+	const { records: searchResults, hasResolved: searchHasResolved } =
+		useEntityRecords( 'taxonomy', taxonomy.slug, {
+			...BASE_QUERY_ARGS,
+			search,
+			orderby: 'name',
+			exclude: termIds,
+			per_page: 20,
+		} );
+
+	suggestionsRef.current = useMemo( () => {
+		if ( ! searchHasResolved ) return suggestionsRef.current;
+
+		const newSuggestions = searchResults.map( ( searchResult: Term ) =>
+			termIdToNameMap.get( searchResult.id )
+		);
+		return newSuggestions;
+	}, [ searchHasResolved, searchResults, termIdToNameMap ] );
+
+	// Update the selected terms when the user selects a suggestion.
+	const onTermsChange = ( newTermValues: FormTokenField.Value[] ) => {
+		const newTermIds = [];
+		for ( const termValue of newTermValues ) {
+			const termId = getTermIdByTermValue(
+				termValue as string | Term,
+				termNameToIdMap
+			);
```

Comment:
I think that if you check whether `termValue` is a string or not here:

```suggestion
			const termId =
				typeof termValue === 'string'
					? getTermIdByTermValue( termValue, termNameToIdMap )
					: termValue.id;
```

Then, `getTermIdByTermValue()` can be a one-liner:

```JS
const getTermIdByTermValue = (
	searchTerm: string,
	termNameToIdMap: Map< string, number >
): number | undefined => {
	return (
		termNameToIdMap.get( searchTerm ) ||
		termNameToIdMap.get( searchTerm.toLocaleLowerCase() )
	);
};
```

Reply:
I understand your point, but I believe it's better to leave the implementation details to our abstraction, i.e. `getTermIdByTermValue`. I have made some improvements to this function based on your feedback in [d9d7cd7](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/d9d7cd77284801aad67c74a2536fc831642acb1d).

However, please let me know if you prefer checking the type outside of `getTermIdByTermValue`. I can make that change because I don't have a strong opinion on this. 🙂

Reply:
Thanks! I usually prefer functions to accept the minimum possible types as possible, so they are easier to understand, but no need to change it.